platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java

```diff
@@ -1124,23 +1126,27 @@ private void writeHashes(final long maxValidPath, final Stream<VirtualHashRecord
     private void writeLeavesToPathToKeyValue(
             final long firstLeafPath,
             final long lastLeafPath,
-            final Stream<VirtualLeafRecord<K, V>> dirtyLeaves,
-            final Stream<VirtualLeafRecord<K, V>> deletedLeaves,
+            @NonNull final Stream<VirtualLeafRecord<K, V>> dirtyLeaves,
+            @NonNull final Stream<VirtualLeafRecord<K, V>> deletedLeaves,
             boolean isReconnect)
             throws IOException {
-        if ((dirtyLeaves == null) || (firstLeafPath <= 0)) {
-            // nothing to do
-            return;
-        }
-
-        // start writing
-        pathToKeyValue.startWriting(firstLeafPath, lastLeafPath);
-        if (!isLongKeyMode) {
-            objectKeyToPath.startWriting();
-        }
+        // If both dirty/deleted leaves streams are empty, no new data files should be created
+        boolean startedWriting = false;
 
         // Iterate over leaf records
-        dirtyLeaves.sorted(Comparator.comparingLong(VirtualLeafRecord::getPath)).forEachOrdered(leafRecord -> {
+        final Iterator<VirtualLeafRecord<K, V>> dirtyIterator = dirtyLeaves
+                .sorted(Comparator.comparingLong(VirtualLeafRecord::getPath))
+                .iterator();
+        while (dirtyIterator.hasNext()) {
```

Comment:
Now that you have iterators you can preserve the previous flow by checking `dirtyIterator.hasNext()` and `deletedIterator.hasNext()` and bailing out if both are false. Will avoid calling `startWriting()` on demand.

Reply:
+1. This way we don't need the `startedWriting` boolean flag here at all.

Reply:
Done